feat: use contrib/depends to build dependencies

MrCyjaneK opened the pull request:
Tested on:
- aarch64-apple-darwin builder:
  - aarch64-apple-darwin (builds + tested)
  - aarch64-apple-ios (builds)
  - x86_64-apple-darwin (builds)
- x86_64-linux-gnu builder:
  - x86_64-unknown-linux-gnu (builds + tested)
  - aarch64-unknown-linux-gnu (builds)
  - x86_64-w64-mingw32 (builds [see note below])
  - aarch64-linux-android (builds)
  - x86_64-linux-android (builds)
  

armv7-linux-androideabi doesn't build due to unrelated issue in unrelated crate (error: failed to run custom build command for `ring v0.17.14`) (simple fixes I tried didn't work or broke unrelated parts of the code so I decided to leave it as it is), essentially it was requiring toolchain that supports api16 - and all other android targets required 21, but it failed way before getting to monero-sys, the buildsystem should work just fine as it follows exactly the same steps as other archs, and uses independent NDK.

**x86_64-pc-windows-gnu** fails with 

```plain
error: couldn't read /workspaces/UnstoppableSwap-core/target/x86_64-pc-windows-gnu/debug/build/crunchy-7e942b50009fa34e/out\lib.rs: No such file or directory (os error 2)
  --> /home/codespace/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/crunchy-0.2.3/src/lib.rs:37:1
   |
37 | include!(concat!(env!("OUT_DIR"), "\\lib.rs"));
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
```

my solution is opening that file and editing \\ -> / - it is unrelated to monero-sys functionality afaik, so I'll leave that up to you. After changing windows abomination into proper POSIX directory separator it builds properly

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Enhanced platform-specific build and bundle configurations, including minimum OS versions for macOS/iOS and improved Windows and iOS build support.
  * Added new submodule and updated existing submodules for Monero integration.
  * Introduced robust error handling for CLI match retrieval in the GUI.
  * Added new dependency for system information gathering.

* **Bug Fixes**
  * Resolved platform-specific build issues, including translation generation, locale inclusion, and iOS-specific build logic.
  * Addressed cross-compilation issues for Linux-to-Windows builds.

* **Chores**
  * Updated and refined build dependencies and scripts for improved reliability and maintainability.
  * Adjusted code formatting exclusions and renamed/expanded cleanup commands.

* **Documentation**
  * Expanded and clarified memory safety and concurrency documentation for the Rust-to-C++ Monero bridge.

* **Style**
  * Updated dependency version specifications for consistency and future compatibility.

* **Configuration**
  * Updated permissions and capabilities for desktop and default environments.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

binarybaron commented:
Cool! I'll try it out in a minute.

binarybaron commented:
bugbot run

MrCyjaneK commented on monero-sys/monero:
```diff
@@ -1 +1 @@
-Subproject commit 5f714f147fd29228698070e6bd80e41ce2f86fb0
+Subproject commit dbbccecc89e1121762a4ad6b531638ece82aa0c7
```
```
$ ls -lah monero-sys/patches/wallet2_api_allow_subtract_from_fee.patch monero-sys/patches/0001-fix-dummy-translation-generator.patch monero-sys/patches/0002-fix-iOS-depends-build.patch monero-sys/patches/0003-include-locale-only-when-targeting-WIN32.patch  
-rw-r--r--@ 1 user  staff   2.2K Jul 14 23:21 monero-sys/patches/0001-fix-dummy-translation-generator.patch
-rw-r--r--@ 1 user  staff   3.7K Jul 14 18:46 monero-sys/patches/0002-fix-iOS-depends-build.patch
-rw-r--r--@ 1 user  staff   1.1K Jul 14 17:59 monero-sys/patches/0003-include-locale-only-when-targeting-WIN32.patch
-rw-r--r--@ 1 user  staff   4.7K Jul 14 16:47 monero-sys/patches/wallet2_api_allow_subtract_from_fee.patch
```

I call bs on this one, files are there.

binarybaron commented:
I'll just start collecting some of my thoughts here. @Einliterflasche will look over this tomorrow as well.

1. Running `cargo build -vv` does not show the output of the build process of the dependencies. It shows us the stdout of `cmake` for the Monero codebase but not for the other dependencies 
2. I have tried compiling `swap` for `aarch64-apple-ios ` with `cargo build --target aarch64-apple-ios -vv --bin=swap`. This currently still fails. A simpler way to reproduce is by building the test in `monero-sys` with `cargo build --target aarch64-apple-ios -vv -p monero-sys --test simple`. This fails due to the following linker error. Some linker errors only get thrown when we actually **use** the functions in our code (I have seen this before)
```
            "___chkstk_darwin", referenced from:
                tools::wallet2::create_transactions_2(std::__1::vector<cryptonote::tx_destination_entry, std::__1::allocator<cryptonote::tx_destination_entry>>, unsigned long, unsigned int, std::__1::vector<unsigned char, std::__1::allocator<unsigned char>> const&, unsigned int, std::__1::set<unsigned int, std::__1::less<unsigned int>, std::__1::allocator<unsigned int>>, std::__1::set<unsigned int, std::__1::less<unsigned int>, std::__1::allocator<unsigned int>> const&) in libmonero_sys-0ba491aacf5b4eeb.rlib[117](wallet2.cpp.o)
                bool do_serialize_container<binary_archive, std::__1::vector<rct::rangeSig, std::__1::allocator<rct::rangeSig>>>(binary_archive<false>&, std::__1::vector<rct::rangeSig, std::__1::allocator<rct::rangeSig>>&) in libmonero_sys-0ba491aacf5b4eeb.rlib[117](wallet2.cpp.o)
                boost::asio::detail::kqueue_reactor::run(long, boost::asio::detail::op_queue<boost::asio::detail::scheduler_operation>&) in libmonero_sys-0ba491aacf5b4eeb.rlib[125](wallet_manager.cpp.o)
                rct::genBorromean(rct::key const*, rct::key const*, rct::key const*, unsigned int const*) in libmonero_sys-0ba491aacf5b4eeb.rlib[221](rctSigs.cpp.o)
                rct::CLSAG_Gen(rct::key const&, std::__1::vector<rct::key, std::__1::allocator<rct::key>> const&, rct::key const&, std::__1::vector<rct::key, std::__1::allocator<rct::key>> const&, rct::key const&, std::__1::vector<rct::key, std::__1::allocator<rct::key>> const&, rct::key const&, unsigned int, hw::device&) in libmonero_sys-0ba491aacf5b4eeb.rlib[221](rctSigs.cpp.o)
                rct::proveRange(rct::key&, rct::key&, unsigned long long const&) in libmonero_sys-0ba491aacf5b4eeb.rlib[221](rctSigs.cpp.o)
                rct::genRct(rct::key const&, std::__1::vector<rct::ctkey, std::__1::allocator<rct::ctkey>> const&, std::__1::vector<rct::key, std::__1::allocator<rct::key>> const&, std::__1::vector<unsigned long long, std::__1::allocator<unsigned long long>> const&, std::__1::vector<std::__1::vector<rct::ctkey, std::__1::allocator<rct::ctkey>>, std::__1::allocator<std::__1::vector<rct::ctkey, std::__1::allocator<rct::ctkey>>>> const&, std::__1::vector<rct::key, std::__1::allocator<rct::key>> const&, unsigned int, std::__1::vector<rct::ctkey, std::__1::allocator<rct::ctkey>>&, rct::RCTConfig const&, hw::device&) in libmonero_sys-0ba491aacf5b4eeb.rlib[221](rctSigs.cpp.o)
                ...
          ld: symbol(s) not found for architecture arm64
          clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

MrCyjaneK commented:
@binarybaron 
> 1. Running cargo build -vv does not show the output of the build process of the dependencies. It shows us the stdout of cmake for the Monero codebase but not for the other dependencies

Fixed.

> 2. I have tried compiling swap for aarch64-apple-ios  with cargo build --target aarch64-apple-ios -vv --bin=swap. This currently still fails. A simpler way to reproduce is by building the test in monero-sys with cargo build --target aarch64-apple-ios -vv -p monero-sys --test simple. This fails due to the following linker error. Some linker errors only get thrown when we actually use the functions in our code (I have seen this before)

Fixed also, but keep in mind that aarch64-apple-darwin target fails if you don't run `cargo clean` right before building it because of patch 0004 failing to apply.

MrCyjaneK commented on monero-sys/build.rs:
```diff
@@ -44,9 +61,75 @@ fn main() {
     // Apply embedded patches before building
     apply_embedded_patches().expect("Failed to apply embedded patches");
 
+    let contrib_depends_dir = std::env::current_dir()
+        .expect("Failed to get current directory")
+        .join("monero_c/contrib/depends");
+
+    let mut target = std::env::var("TARGET").unwrap_or_else(|_| "unknown".to_string());
+    if target == "aarch64-unknown-linux-gnu" {
+        target = "aarch64-linux-gnu".to_string();
+    }
+    if target == "armv7-linux-androideabi" {
+        target = "armv7a-linux-androideabi".to_string();
+    }
+    if target == "x86_64-pc-windows-gnu" {
+        target = "x86_64-w64-mingw32".to_string();
+    }
+    println!("cargo:warning=Building for target: {}", target);
```
@coderabbitai so smart yet can't use github suggestion so I could just press [commit suggestion] smh
